feat(preprod): Add preprod images download endpoint

rbro112 opened the pull request:
Adds a new preprod `/files/images/<image_id>` endpoint which allows our frontend to download images from an ID with objectstore. First to be used with app icons coming soon.

Tested fully E2E locally and all works well. To work, this is reliant on landing https://github.com/getsentry/launchpad/pull/430 once we get a published version of the objectstore client, but since nobody will be consuming this endpoint until we land the stacked PR (#102118) this is safe to go ahead and merge.

rbro112 commented:
* **#102118** <a href="https://app.graphite.dev/github/pr/getsentry/sentry/102118?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#102117** <a href="https://app.graphite.dev/github/pr/getsentry/sentry/102117?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/getsentry/sentry/102117?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

rbro112 commented on src/sentry/preprod/analytics.py:
```diff
@@ -77,6 +77,23 @@ class PreprodArtifactApiDeleteEvent(analytics.Event):
 
 
+@analytics.eventclass("preprod_artifact.api.image")
+class PreprodArtifactApiImageEvent(analytics.Event):
```
This is the new analytics event

trevor-e commented on src/sentry/preprod/api/endpoints/project_preprod_artifact_image.py:
```diff
@@ -0,0 +1,148 @@
+        try:
+            result = client.get(object_key)
+            # Read the entire stream at once
+            image_data = result.payload.read()
```
nit: is this necessary or is there a way to forward to stream to the HttpResponse for handling?

noahsmartin commented on src/sentry/preprod/api/endpoints/project_preprod_artifact_image.py:
```diff
@@ -0,0 +1,116 @@
+logger = logging.getLogger(__name__)
+
+
+def detect_image_content_type(image_data: bytes) -> str:
```
From Jan's comment below it sounds like eventually this can be set on write and we wouldn't have to do it here on read, but this seems ok for now to unblock testing

rbro112 commented on src/sentry/preprod/api/models/project_preprod_build_details_models.py:
```diff
@@ -37,6 +37,7 @@ class BuildDetailsAppInfo(BaseModel):
     platform: Platform | None = None
     is_installable: bool
     build_configuration: str | None = None
+    app_icon_id: str | None = None
```
This is related to app icon, not the generic download endpoint, but should be fine to merge as the frontend does not yet use this until https://github.com/getsentry/sentry/pull/102118 lands

Abdkhan14 committed:
feat(explore-attr-breakdowns): Adding chart selection CTA (#103652)

<img width="1419" height="682" alt="Screenshot 2025-11-19 at 11 20
18 AM"
src="https://github.com/user-attachments/assets/80d5eb1e-058b-49f0-b8e0-ab4cc833835c"
/>

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

rbro112 commented on src/sentry/objectstore/__init__.py:
```diff
@@ -49,3 +52,17 @@ def get_attachments_session(org: int, project: int) -> Session:
     return _ATTACHMENTS_CLIENT.session(_ATTACHMENTS_USECASE, org=org, project=project)
+
+
+def get_preprod_session(org: int, project: int) -> Session:
```
Will do this in a follow up, want to minimize risk for us here so I'd prefer not to touch existing _ATTACHMENTS_CLIENT

jan-auer replied:
We should default to using just a single client for all use cases. If you want to play it extra safe in the beginning, that's fine, but eventually I would suggest to maintain one `_OBJECTSTORE_CLIENT` and simply create separate sessions using that.

jan-auer commented on src/sentry/objectstore/__init__.py:
```diff
@@ -40,6 +40,9 @@ def distribution(
 _ATTACHMENTS_CLIENT: Client | None = None
 _ATTACHMENTS_USECASE = Usecase("attachments", expiration_policy=TimeToLive(timedelta(days=30)))
 
+_PREPROD_CLIENT: Client | None = None
```
Now that we have multiple use cases, I think we should default to reusing the same client for all of them. Mind updating this so that both attachments and preprod use a single `_OBJECTSTORE_CLIENT` or similar?

jan-auer commented on src/sentry/preprod/api/endpoints/project_preprod_artifact_image.py:
```diff
@@ -0,0 +1,55 @@
+        try:
+            result = session.get(object_key)
+            # Read the entire stream at once (necessary for content_type)
+            image_data = result.payload.read()
```
We have a generic endpoint that proxies through to objectstore now (`OrganizationObjectstoreEndpoint`). What if you issue a temporary redirect (307) instead and let that endpoint do the proxying? This will also include the content-type header.

@lcian we should add a `get_proxy_url` to `sentry.objectstore` similar to `get_symbolicator_url` that outputs the route. That allows us to add the signature once we ship it.

rbro112 replied:
Also we can just leverage the generic endpoint from our frontend use-case, but also will suffer from the same limitations of the feature flag here, so once that's enabled for a wider audience I can make that change and just likely remove this endpoint entirely.

rbro112 committed:
Feedback
